src/main/java/org/apache/maven/plugins/toolchain/jdk/ToolchainDiscoverer.java

```diff
@@ -0,0 +1,433 @@
+        Stream.of(PROPERTIES).forEach(k -> {
+            String v = System.getProperty(JAVA + k);
+            if (v != null) {
+                model.addProvide(k.substring(JAVA.length()), v);
+            }
+        });
+        model.addProvide(CURRENT, "true");
+        Xpp3Dom config = new Xpp3Dom("configuration");
+        Xpp3Dom jdkHome = new Xpp3Dom(JDK_HOME);
+        jdkHome.setValue(currentJdkHome.toString());
+        config.addChild(jdkHome);
+        model.setConfiguration(config);
+        return Optional.of(model);
+    }
+
+    public PersistedToolchains discoverToolchains() {
+        return discoverToolchains(LTS + COMMA + VERSION + COMMA + VENDOR);
+    }
+
+    /**
+     * Returns a PersistedToolchains object containing a list of discovered toolchains,
+     * never <code>null</code>.
+     */
+    public PersistedToolchains discoverToolchains(String comparator) {
+        try {
+            Set<Path> jdks = findJdks();
+            log.info("Found " + jdks.size() + " possible jdks: " + jdks);
+            cacheModified = false;
+            readCache();
+            Map<Path, Map<String, String>> flags = new HashMap<>();
+            Path currentJdkHome = getCanonicalPath(Paths.get(System.getProperty(JAVA_HOME)));
+            flags.computeIfAbsent(currentJdkHome, p -> new HashMap<>()).put(CURRENT, "true");
+            // check environment variables for JAVA{xx}_HOME
+            System.getenv().entrySet().stream()
+                    .filter(e -> e.getKey().startsWith("JAVA") && e.getKey().endsWith("_HOME"))
+                    .forEach(e -> {
+                        Path path = getCanonicalPath(Paths.get(e.getValue()));
+                        Map<String, String> f = flags.computeIfAbsent(path, p -> new HashMap<>());
+                        String val = f.getOrDefault(ENV, "");
+                        f.put(ENV, (val.isEmpty() ? "" : val + ",") + e.getKey());
+                    });
+
+            List<ToolchainModel> tcs = jdks.parallelStream()
+                    .map(s -> {
+                        ToolchainModel tc = getToolchainModel(s);
+                        flags.getOrDefault(s, Collections.emptyMap())
+                                .forEach((k, v) -> tc.getProvides().setProperty(k, v));
+                        String version = tc.getProvides().getProperty(VERSION);
+                        if (isLts(version)) {
+                            tc.getProvides().setProperty(LTS, "true");
+                        }
+                        return tc;
+                    })
+                    .sorted(getToolchainModelComparator(comparator))
+                    .collect(Collectors.toList());
+            if (this.cacheModified) {
+                writeCache();
+            }
+            PersistedToolchains ps = new PersistedToolchains();
+            ps.setToolchains(tcs);
+            return ps;
+        } catch (Exception e) {
+            if (log.isDebugEnabled()) {
```

Comment:
My goal is the following: the toolchain discovery is optional, so any exception happening should be logged to the user but not forbid the use of the feature, in particular a later selection of the toolchain.  Any exception caught is logged at WARNING level (with stack trace if debug is enabled). 
This can be moved to the callers, though...